text/0000-spec-requirements.md

```diff
@@ -0,0 +1,53 @@
+# Spec Requirements
+
+## Test writing
+
+### Tests are written as framework-agnostic
+* Should support unit and end-to-end tests
+* JavaScript executed on the DOM is separate from the assertions
```

Comment:
@scottnath I believe this hits at an implementation detail rather than a requirement. There may be ways to build systems now, or in the future, such that they would not need to separate these out while still meeting the underlying requirement. I think we should define the underlying requirement and refer to that instead of specifying how the system itself should be built.

Reply:
The requirement is meant to convey that the DOM manipulation, which _can_ be framework-specific, should be separate from the assertions, which should never be framework-specific. 

trying to figure out how to write the requirement correctly, so the below examples show one benefit of separate DOM manipulation from the test

## Button HTML
**React**
```html
<button tabindex="0" class="some-class bx--btn bx--btn--primary" type="button">
  Button
  <svg focusable="false" preserveAspectRatio="xMidYMid meet" aria-label="Search" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" role="img" class="bx--btn__icon" style="will-change: transform;"><path d="M15 14.3L10.7 10c1.9-2.3 1.6-5.8-.7-7.7S4.2.7 2.3 3 .7 8.8 3 10.7c2 1.7 5 1.7 7 0l4.3 4.3.7-.7zM2 6.5C2 4 4 2 6.5 2S11 4 11 6.5 9 11 6.5 11 2 9 2 6.5z"></path></svg>
</button>
```

## DOM manipulation and assertion in same function in the test


```js
const rules = [
  {
    id: 'button.exists',
    description:
      'Single button with icon',
    level: 'error',
    context: defaultContext,
    test: window => {
      const button = window.document.querySelector('.bx--button');
      const icon = button.querySelector(selectors.icon);

      expect(button).to.not.exist;
      expect(icon).to.not.exist;
    }
  },
  {
    id: 'buttons.exist',
    description:
      'Multiple buttons with icons',
    level: 'error',
    context: defaultContext,
    test: window => {
      const buttons = window.querySelectorAll('.bx--btn');

      for (let i = 0; i < buttons.length; i++) {
        // these are the exact same assertions as above
        expect(button[i]).to.exist;
        expect(icon[i]).to.exist;
      }
    }
  }
};

module.exports = rules;
```

### Using the test

```js

describe('Testing button component', () => {
  rules.forEach(rule => {
    it(rule.description, () => {
      const window = _getsWindowFromContext(rule.context);
      rule.test(window)
    })
  })
});
```

## DOM manipulation and assertion separated

### Test setup

```js
const tests = [
  {
    description: 'Button and icon exist',
    getResults: window => {
      const button = window.document.querySelector('.bx--button');
      const icon = button.querySelector('.bx--icon');

      return {
        button,
        icon
      }
    },
    comparison: results => {
      expect(results.button).to.exist;
      expect(results.icon).to.exist;
    }
  }
]

module.exports = tests;
```


### Using the test

(note: this is pseudo-code, in reality the end user wouldn't use `rule[0]` but would instead be either iterating through the tests, or with the requirement below for _addressable_ test groups, the user could iterate through just the tests in one group)

```js

describe('Testing button component', () => {
  it(rule[0].description, () => {
    const window = _getsWindowFromContext('find-single-button-page');
    const testContent = rule[0].getResults(window)
    rule.comparison(testContent)
  })
  it('Testing multiple buttons with icons exist', () => {
    // note: this example doesn't use `getResults` at all, allowing the functionality of 
    //    getting the content to test to be overwritten by the end-user
    const window = _getsWindowFromContext('find-multiple-button-page');
    const buttons = window.document.querySelector('.bx--btn');
    for (let i = 0; i < buttons.length; i++) {
      const testContent = {
        button: buttons[i],
        icon: buttons[i].querySelector('.bx--icon')
      }

      // this allows the requirements for a single button to come from a single place
      //    which also makes it easier to change a component because it would have
      //    a common set of tests which could be used in many different situations
      rule.comparison(testContent)
    }
  })
});
```